src/dune_rules/gen_rules.ml

```diff
@@ -374,6 +374,35 @@ let has_rules subdirs f =
        ; rules
        })
 
+let is_under_melange_emit dir =
+  let all_parents dir =
+    let rec inner acc dir =
+      match Path.Build.parent dir with
+      | None -> acc
+      | Some dir -> inner (dir :: acc) dir
+    in
+    inner [] dir
+  in
+
+  (* Walk up the parent directories to check if one of them declares a
+     `melange.emit` stanza *)
+  let rec does_parent_have_melange_emit = function
+    | [] -> Memo.return false
+    | dir :: rest -> (
+      Only_packages.stanzas_in_dir dir >>= function
+      | None -> does_parent_have_melange_emit rest
+      | Some { Dune_file.stanzas; dir = _; project = _ } -> (
+        match
+          List.exists stanzas ~f:(function
+            | Melange_emit _ -> true
+            | _ -> false)
+        with
+        | true -> Memo.return true
+        | false -> does_parent_have_melange_emit rest))
+  in
+
+  does_parent_have_melange_emit (List.rev (all_parents dir))
```

Comment:
Simpler would be to pass the directory directly and call `Path.Build.parent` until it returns `None`.

Reply:
That is, in fact, a lot better. done in [e9f9966](https://github.com/ocaml/dune/pull/6579/commits/e9f9966255f3677a61cd526cef0d43ca31889ae6)